### PR TITLE
Add usePreventWindowUnload hook.

### LINKED ISSUE #799: Show an error message if user leaves editor without saving
## Bug Description

If a user leaves the editor without saving the story, all progress / changes will be lost. 

## Expected Behaviour

Some kind of error message would be shown to user if they try and navigate away.

## Steps to Reproduce

- Create a story
- Add some elements.
- Click on any link in the left hand sidebar.

## Screenshots

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### LINKED ISSUE #865: Users can reload or leave the window before upload finish.
## Bug Description

Users can reload or leave the window before an upload has finished without any warning.

## Expected Behaviour

As described in https://github.com/google/web-stories-wp/issues/231#issuecomment-605947877, the user should receive a warning prompt if try to redirect the current editor page or reload. This has the same behavior as described in #799, but for `uploading` state.

## Steps to Reproduce

1 - Create a Story.
2 - Drop a file to the story workspace or media library created in the step 1.
3 - Reload the page without wait the upload finish (or redirect the browser URL to another URL and back to the story created in step 1)
4 - When the story created in the step 1 loads again, it will not retrieve the upload file, since it was force canceled in step 3.

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by obetomuniz**
Fixes #799 
Fixes #865 

Kapture-2020-03-31-at-14 49 07

### CODE REVIEW

**Comment by spacedmonkey** [comment]
In my testing, if you open the editor do nothing and try and leave, I still see a message saying I will leave my changes. This seems weird, as I haven't done anything yet.

**Comment by obetomuniz** [comment]
Interesting @spacedmonkey. It should not happen since it is using the changes history to control when to register the event. I will take a look for potential edge cases.

**Comment by spacedmonkey** [comment]
@obetomuniz Here is a video of it happening - https://youtu.be/LocZ6bR4EGY

My money is that something is trigger an event that goes into the history. Seems like the undo button is highlighted, so something happened. But easy to replicate.

File: assets/src/edit-story/utils/usePreventWindowUnload.js
Comment (dvoytenko) [inline assets/src/edit-story/utils/usePreventWindowUnload.js]: s/Object/Event

File: assets/src/edit-story/utils/usePreventWindowUnload.js
Comment (dvoytenko) [inline assets/src/edit-story/utils/usePreventWindowUnload.js]: Why are there two `removeEventListener` with the same args: here and in the cleanup. It seems that the one here is not needed?

**Comment by obetomuniz** [comment]
Perfect example @spacedmonkey . Thanks for the video, it will help attack the case.

File: assets/src/edit-story/utils/usePreventWindowUnload.js
Comment (dvoytenko) [inline assets/src/edit-story/utils/usePreventWindowUnload.js]: The way it's written, this is a rather dangerous hook to be run wherever. It should only be ever invoked in a minimal number of places (ideally only one place). One option is to make this code private to either `HistoryProvider`, or `useSaveStory`, or maybe `StoryProvider`. And "has changes" signal somehow calculated there or "saved" signal is propagated from `useSaveStory`. Generally, such a singleton task is better suited for a provider/private hook than a reusable hook.

File: assets/src/edit-story/app/history/historyProvider.js
Comment (dvoytenko) [inline assets/src/edit-story/app/history/historyProvider.js]: The `canUndo` survives editor savings. Thus, I don't think it can be used here - it will be regularly overwriting the `usePreventWindowUnload(false)` set in the save story.

**Review by dvoytenko — CHANGES_REQUESTED** [review]

File: assets/src/edit-story/utils/usePreventWindowUnload.js
Comment (obetomuniz) [reply assets/src/edit-story/utils/usePreventWindowUnload.js]: Initially, I created something under the HistoryProvider scope, but only for reusability of the logic, I moved to a hook since we will need to control this prompt in more places (For instance, uploads not completed), not only during the save. Anyway, I will take a look at your insights and move it back to some already existing stuff that controls the history and/or saves actions.

File: assets/src/edit-story/app/history/historyProvider.js
Comment (obetomuniz) [reply assets/src/edit-story/app/history/historyProvider.js]: It only will override the `usePreventWindowUnload(false)` set in the save story action when you have something new in history. But, as @spacedmonkey reported, there are specific cases that we have "hidden" history changes that can in fact create side effects. 

That being said, I am a bit in doubt about how to track truly canvas changes since the history is not trustable. Maybe have a dedicated action for it? Do you guys have another suggestion?

Comment (dvoytenko) [reply assets/src/edit-story/app/history/historyProvider.js]: I think what you're saying, if the case, is a separate issue for history where it should not have "hidden" updates. However, my point above still stands `canUndo` will almost always be `true`: including for a just saved story. That means, at any random time, this code might execute `usePreventWindowUnload(true)`, including just after saving a story. What we want to do here, `usePreventWindowUnload(hasChangedSinceLastSave)`. If we had it here, you wouldn't need to call `usePreventWindowUnload` from anywhere else.

Comment (obetomuniz) [reply assets/src/edit-story/app/history/historyProvider.js]: Sure! I'm working on it. Just not sure yet from where I will extract a trustable history changes. My best guess is still `historyLength` :)

File: assets/src/edit-story/utils/usePreventWindowUnload.js
Comment (barklund) [inline assets/src/edit-story/utils/usePreventWindowUnload.js]: So this is only exported for testing purposes? You could just test for `expect.any(Function)` instead - I think that's acceptable under these circumstances.

File: assets/src/edit-story/utils/usePreventWindowUnload.js
Comment (barklund) [reply assets/src/edit-story/utils/usePreventWindowUnload.js]: We just need it to be extensible by e.g. uploads in progress and similar - is that still doable?

File: assets/src/edit-story/app/history/historyProvider.js
Comment (barklund) [inline assets/src/edit-story/app/history/historyProvider.js]: So, this will only run when these values change. But `canUndo` will stay the same after the first change after a save, no? So, making a change after saving will not trigger here? I think a better test would be to make sure it runs again for every change to the history (though saving will update the story properties which will in turn cause a change in history?).

I guess my question is: Does this actually work?

**Review by barklund — COMMENTED** [review]
A couple of questions

File: assets/src/edit-story/app/history/historyProvider.js
Comment (obetomuniz) [reply assets/src/edit-story/app/history/historyProvider.js]: Just fixed, we should store the records for reference. Thanks for point this out.

File: assets/src/edit-story/app/history/historyProvider.js
Comment (barklund) [inline assets/src/edit-story/app/history/historyProvider.js]: Note that when history reaches its buffer size, it'll prune from the other end and just stay at the same length continuously. Does this still work with a full history buffer?

Comment (barklund) [reply assets/src/edit-story/app/history/historyProvider.js]: The length is [currently set to only 50](https://github.com/google/web-stories-wp/blob/bug/alert-user-before-exit-window/assets/src/edit-story/app/index.js#L54), so it'll be reached fairly quickly.

Maybe we want another flag coming from the history reducer with a _"history version number"_ or something like that - a number that increments when ever history changes.

@dvoytenko, does that make sense to you?

Comment (obetomuniz) [reply assets/src/edit-story/app/history/historyProvider.js]: Indeed. I also figured this problem after your first comment. I just pushed a proposal (which is basically what you said in order to create a global history length counter). Also, thanks for the report, I did not note this buffer capability in the history changes while creating this.

File: assets/src/edit-story/utils/usePreventWindowUnload.js
Comment (obetomuniz) [reply assets/src/edit-story/utils/usePreventWindowUnload.js]: Yup.

File: assets/src/edit-story/app/history/historyProvider.js
Comment (spacedmonkey) [inline assets/src/edit-story/app/history/historyProvider.js]: A comment to explain what is going on here would be useful.

File: assets/src/edit-story/app/history/historyProvider.js
Comment (spacedmonkey) [inline assets/src/edit-story/app/history/historyProvider.js]: Shouldnt this be part of the history reducer?

Comment (spacedmonkey) [reply assets/src/edit-story/app/history/historyProvider.js]: This feel weird to have a reducer and state next to each other like this.

**Comment by spacedmonkey** [comment]
I have created a related issue for adding this hook while files are uploading - https://github.com/google/web-stories-wp/issues/971

File: assets/src/edit-story/app/history/useHistoryReducer.js
Comment (spacedmonkey) [inline assets/src/edit-story/app/history/useHistoryReducer.js]: Should the global history remove one if you undo?

File: assets/src/edit-story/utils/usePreventWindowUnload.js
Comment (spacedmonkey) [reply assets/src/edit-story/utils/usePreventWindowUnload.js]: Created a ticket for upload files - https://github.com/google/web-stories-wp/issues/971

File: assets/src/edit-story/app/history/useHistoryReducer.js
Comment (barklund) [inline assets/src/edit-story/app/history/useHistoryReducer.js]: I'm not sure this makes sense. Isn't the purpose simply to track, that "something" has changed?

I'd say this variable should always increase. And maybe rename it to `versionNumber` or `historyID` even, as it's not really useful as an actual value (shouldn't be used to say anything about the actual length of history nor should the difference between two values mean anything), just as a number, that "changes" (to something new) when a user does "anything".

Comment (obetomuniz) [reply assets/src/edit-story/app/history/useHistoryReducer.js]: Good point @barklund, but @spacedmonkey suggested https://github.com/google/web-stories-wp/pull/877#discussion_r402223355 track `undo` changes (For instance, if the user `undo` until the initial state, the prevent window unload should be disabled, and if the user `redo` to some point ahead, it should be enabled again), which makes sense for me, or I misunderstood that? 

From this point, what should I do? Do you have another way to propose track `undo` if not decreasing that value? I agree about rename to `versionNumber` or `historyID`. I liked both, but since it changes progressively, I would use `versionNumber.` If we decide to move forward tracking the `undo`/`redo` actions, I would prefer `historyID`.

File: assets/src/edit-story/app/history/historyProvider.js
Comment (dvoytenko) [reply assets/src/edit-story/app/history/historyProvider.js]: Indeed, `historyLength` is definitely a buffer thing. We need `updateCounter` or `versionId` or something that keeps incrementing forever in the editor's instance.

File: assets/src/edit-story/app/history/historyProvider.js
Comment (dvoytenko) [inline assets/src/edit-story/app/history/historyProvider.js]: Are all these expressions simple `setHistoryChangedState(globalHistoryLEngth - 1 > 0)`?

File: assets/src/edit-story/app/history/useHistoryReducer.js
Comment (dvoytenko) [reply assets/src/edit-story/app/history/useHistoryReducer.js]: I think it should just keep incrementing. Maybe some name would clarify this better. But it's a new version no matter what.

File: assets/src/edit-story/app/history/useHistoryReducer.js
Comment (dvoytenko) [inline assets/src/edit-story/app/history/useHistoryReducer.js]: Why do we need to have both `hasHistoryChanged` and `globalHistoyLength`?

File: assets/src/edit-story/app/history/useHistoryReducer.js
Comment (dvoytenko) [reply assets/src/edit-story/app/history/useHistoryReducer.js]: I agree. I think we just need one ever-incrementing value. It could be either truly global or it could be "number of changes since the last save". Either would solve the problem for us.

**Review by dvoytenko — CHANGES_REQUESTED** [review]

File: assets/src/edit-story/app/history/useHistoryReducer.js
Comment (spacedmonkey) [reply assets/src/edit-story/app/history/useHistoryReducer.js]: I think a user should be able to exit an empty story without hitting this message. What if I create a story, create an element, click undo and try and exit. Global history will be two, as element then undoing that element. This just feels a little weird.

File: assets/src/edit-story/app/history/useHistoryReducer.js
Comment (spacedmonkey) [reply assets/src/edit-story/app/history/useHistoryReducer.js]: `hasHistoryChanged` is just `globalHistoyLength > 0`

File: assets/src/edit-story/app/history/useHistoryReducer.js
Comment (obetomuniz) [reply assets/src/edit-story/app/history/useHistoryReducer.js]: Sure, so for now, I will revert the undo/redo logic and only track registers after save/load events. If later we decide to add this UX improvement, we can revert it again.

File: assets/src/edit-story/app/history/useHistoryReducer.js
Comment (obetomuniz) [reply assets/src/edit-story/app/history/useHistoryReducer.js]: Actually not, `hasHistoryChanged` allows to control explicitly a temporary scope of history changes, avoiding race conditions registering `beforeunload` event. For this scenario, only a counter is not enough, we also should consider that our system update the versionNumber systematically without user intention (e.g. creating a new story, uploading files, etc)

File: assets/src/edit-story/app/history/historyProvider.js
Comment (dvoytenko) [inline assets/src/edit-story/app/history/historyProvider.js]: This seems recursive. If reducer is responsible for both `versionNumber` and `hasHistoryChanged`, then why do we need to update it in the reducer?

Also, I still find the amount of exposed data too much here. It seems all we need to return from here is whether or not the "document has changed since the last save".  It's a single boolean value (`hasChangedSinceLastSave`) and one action function (`storySaved()`). The `versionNumber`, etc - all could just be internal details and probably can be completely baked inside the reducer?

File: assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js
Comment (dvoytenko) [inline assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js]: I don't necessarily see that the uploading has much to do with history itself. I think we're trying to hack away here from a somewhat different underlying structure that involves a hook and a provider. I think trying to skip the provider, makes it harder for us here. Let's discuss something like this:

/1/ There's a high-level `PreventWindowUnloadProvider`. It it exposes the action:
```
const setPreventUnload = usePreventWindowUnloadHook();
```

/2/  The history work is all the same and it calls:
```
const setPreventUnload = usePreventWindowUnloadHook();
useEffect(() => {
  setPreventUnload('history', hasChangedSinceLastSave);
}, [hasChangedSinceLastSave]);
```

/3/ The uploader is about the same:
```
const setPreventUnload = usePreventWindowUnloadHook();
// We can either use a single "upload" ID, or a unique ID per uploading session.
setPreventUnload('upload', true);
uploadFile(...).finally(() => setPreventUnload('upload', false));
```

/4/ The `PreventWindowUnloadProvider` is thus pretty simple:
```
function PreventWindowUnloadProvider() {
  const lockRef = useRef(new Set());
  const setPreventUnload = useCallback((id, value) => {
    if (value) {
      lockRef.current.add(id);
    } else {
      lockRef.current.delete(id);
    }
  }, []);
  useEffect(() => {
    window.addEventListener(..., (evt) => {
      if (lockRef.current.size > 0) {
        evt.preventDefault();
        // ....
      }
    });
    return () => window.removeEventListener(...);
  }, []);
}
```

**Review by dvoytenko — CHANGES_REQUESTED** [review]

File: assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js
Comment (obetomuniz) [reply assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js]: I really liked this structure. Yeah, let me apply something on this way, and check if the side effects stay affecting, but it makes much sense to me. Thanks @dvoytenko

File: assets/src/edit-story/app/history/historyProvider.js
Comment (dvoytenko) [inline assets/src/edit-story/app/history/historyProvider.js:46]: One more important nuance here possible. You also need an additional unmount effect:

```
useEffect(() => {
  return () => setPreventUnload('history', false);
}, [setPreventUpload])
```

This would be the most precise way to unblock unloading even if the component was unmounted.

File: assets/src/edit-story/app/story/effects/usePageBackgrounds.js
Comment (dvoytenko) [inline assets/src/edit-story/app/story/effects/usePageBackgrounds.js]: I really don't understand why history is cleared here. Could you clarify?

File: assets/src/edit-story/utils/usePreventWindowUnload.js
Comment (dvoytenko) [inline assets/src/edit-story/utils/usePreventWindowUnload.js]: You mean `usePreventWindowUnload` here? Since it's not a provider?

File: assets/src/edit-story/utils/usePreventWindowUnload.js
Comment (dvoytenko) [inline assets/src/edit-story/utils/usePreventWindowUnload.js]: This will work but with side-effects. IMHO it'd be a lot better to use `createContext()` here with the default value that will have what you want. Then you can use `useContext()` inside your `usePreventWindowUnload()`. You won't need any additional extra provider declarations - default context will work exactly like you need here. That's exactly what it's for.

E.g.

```
const PreventUnloadContext = createContext({listeners: new Map()});

function usePreventWindowUnload(...) {
  const context = useContext(PreventUnloadContext);
  const setPreventUnload = useCallback(() => {
    context.listeners.set(...);
     ...
  });
  return setPreventUnload;
}
```

**Review by dvoytenko — CHANGES_REQUESTED** [review]
This looks pretty much there, but one big question related to clearing of history in background work.

File: assets/src/edit-story/app/story/effects/usePageBackgrounds.js
Comment (obetomuniz) [reply assets/src/edit-story/app/story/effects/usePageBackgrounds.js]: Right above we have `createNewElement`, `addElement`, and `setBackgroundElement` that we use to create a white rectangle for new stories. It is incrementing an unnecessary new history version, which also enables the prevent window unload. Adding this `clearHistory` call resets the history and ensure that starting new stories will not create any fake history change only because we add a fake (since its an element) white background on canvas.

Comment (obetomuniz) [reply assets/src/edit-story/app/story/effects/usePageBackgrounds.js]: If we remove this, the user will always receive a prevent window unload alert if it tries to leave the page of a new story, even without direct changes.

Comment (barklund) [reply assets/src/edit-story/app/story/effects/usePageBackgrounds.js]: But this also means, that adding a page and pressing reload doesn't warn the user, even though you've clearly made a change.

It really only should be the very first page added to the story, that should be ignored for these purposes - not subsequent pages.

Also not that @miina is changing the logic of the page background in #1129 .

Comment (obetomuniz) [reply assets/src/edit-story/app/story/effects/usePageBackgrounds.js]: Since we are adding an element (for new stories/pages), we should expect that a new history state will be handled. Should I add an independent `new-page` listener to prevent window unload when the user handles the action to add a New Page and remove it as soon as the user Save the story? So we can stay with this clearHistory without problems.

Comment (dvoytenko) [reply assets/src/edit-story/app/story/effects/usePageBackgrounds.js]: I'd rather prevent unload was slightly broken than workaround this issue this way. Let's remove `clearHistory` from here for now.

Comment (dvoytenko) [reply assets/src/edit-story/app/story/effects/usePageBackgrounds.js]: No. I don't think this is the right place to solve this problem. I think the issue with the new-page will have to be solved separately. From the point of view of `usePreventWindowUnload` as a user of the history system, it should be reasonable to demand that the history is "clean".

**Review by dvoytenko — CHANGES_REQUESTED** [review]
I'm ready to approve this as soon as we remove `clearHistory`.

File: assets/src/edit-story/app/story/effects/usePageBackgrounds.js
Comment (obetomuniz) [reply assets/src/edit-story/app/story/effects/usePageBackgrounds.js]: Exactly. Also, new stories without any user intentional changes can't create history state updates. That would fix the problem. Anyway, let me remove `clearHistory` for now. Thanks @dvoytenko and @barklund for the insights here.

**Comment by obetomuniz** [comment]
Done @dvoytenko 👍

**Review by dvoytenko — APPROVED** [review]

File: assets/src/edit-story/app/story/effects/usePageBackgrounds.js
Comment (barklund) [reply assets/src/edit-story/app/story/effects/usePageBackgrounds.js]: With @miina's latest solution in #1129 this might actually be fixed magically, as the [`restore` in `useLoadStory`](https://github.com/google/web-stories-wp/blob/master/assets/src/edit-story/app/story/effects/useLoadStory.js#L127-L133) will already have the background element present ([as it will be added in `createPage`](https://github.com/google/web-stories-wp/blob/master/assets/src/edit-story/app/story/effects/useLoadStory.js#L91)) and thus not cause a history change.

Do however make a note to check this (a new issue perhaps) once both these issues are merged. If it's still a problem then, we'll deal with it later.

Comment (obetomuniz) [reply assets/src/edit-story/app/story/effects/usePageBackgrounds.js]: Just issued on #1137.

Comment (miina) [reply assets/src/edit-story/app/story/effects/usePageBackgrounds.js]: Correct, #1129 will address that, too.

**Comment by dvoytenko** [comment]
LGTM

**Review by barklund — APPROVED** [review]

**Comment by barklund** [comment]
The [failing test](https://travis-ci.com/github/google/web-stories-wp/jobs/317856292#L631) should be fixed by merging in master (as #1129 has been merged), as it actually specifically fails because you can't add `onBeforeUnload` before any user interaction.

Didn't know you couldn't do that, TIL!

**Comment by barklund** [comment]
@googlebot I consent.

**Pull request merged**
